src/url.cpp

```diff
@@ -0,0 +1,116 @@
+#include <algorithm>
+#include <string>
+#include <iterator>
+#include <unordered_map>
+#include <unordered_set>
+#include <iostream>
+
+#include "url.h"
+
+namespace Url {
+
+    /* Character classes */
+    const std::string Url::GEN_DELIMS = ":/?#[]@";
+    const std::string Url::SUB_DELIMS = "!$&'()*+,;=";
+    const std::string Url::ALPHA = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    const std::string Url::DIGIT = "0123456789";
+    const std::string Url::UNRESERVED = Url::ALPHA + Url::DIGIT + "-._~";
+    const std::string Url::RESERVED = Url::GEN_DELIMS + Url::SUB_DELIMS;
+    const std::string Url::PCHAR = Url::UNRESERVED + Url::SUB_DELIMS + ":@";
+    const std::string Url::PATH = Url::PCHAR + "/";
+    const std::string Url::QUERY = Url::PCHAR + "/?";
+    const std::string Url::FRAGMENT = Url::PCHAR + "/?";
+    const std::string Url::USERINFO = Url::UNRESERVED + Url::SUB_DELIMS + ":";
+    const std::string Url::SCHEME =
+        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+-.";
+
+    Url::Url(const std::string& url): port(0) {
+        std::string copy(url);
+
+        size_t index = copy.find(':');
```

Comment:
Is the plan for this code to do a literal translation first and then optimize? I'm fine if that's the plan, but my first pass would be to avoid the multiple copies and reassignments.